mreg_cli/api/models.py

```diff
@@ -386,8 +399,400 @@ def is_delegated(self) -> bool:
+        """
+        if not name.endswith(f".{self.name}"):
+            raise InputFailure(f"Delegation '{name}' is not in '{self.name}'")
+
+    @classmethod
+    def type_by_name(cls, name: str) -> type[ForwardZone | ReverseZone]:
+        """Determine the zone type based on the name.
+
+        :param name: The name of the zone.
+        :returns: The zone type.
+        """
+        if is_reverse_zone_name(name):
+            return ReverseZone
+        return ForwardZone
+
+    @classmethod
+    def verify_nameservers(cls, nameservers: list[str], force: bool = False) -> None:
+        """Verify that nameservers are in mreg and have A-records."""
+        if not nameservers:
+            raise InputFailure("At least one nameserver is required")
+
+        errors: list[str] = []
+        for nameserver in nameservers:
+            try:
+                host = Host.get_by_any_means_or_raise(nameserver)
+            except EntityNotFound:
+                if not force:
+                    errors.append(f"{nameserver} is not in mreg, must force")
+            else:
+                if host.zone is None:
+                    if not host.ipaddresses and not force:
+                        errors.append(f"{nameserver} has no A-record/glue, must force")
+        if errors:
+            raise CliWarning("\n".join(errors))
+
+    @classmethod
+    def create_zone(
+        cls,
+        name: str,
+        email: str,
+        primary_ns: list[str],
+        force: bool,
+    ) -> ForwardZone | ReverseZone | None:
+        """Create a forward or reverse zone based on zone name.
+
+        :param name: The name of the zone to create.
+        :param email: The email address for the zone.
+        :param primary_ns: The primary nameserver for the zone.
+        :returns: The created zone object.
+        """
+        cls.verify_nameservers(primary_ns, force=force)
+        zone_t = cls.type_by_name(name)
+        zone_t.get_zone_and_raise(name)
+        return zone_t.create({"name": name, "email": email, "primary_ns": primary_ns})
+
+    @classmethod
+    def get_zone(cls, name: str) -> ForwardZone | ReverseZone | None:
+        """Get a zone by name.
+
+        :param name: The name of the zone to get.
+        :returns: The zone object.
+        """
+        zone_t = cls.type_by_name(name)
+        return zone_t.get_by_name(name)
+
+    @classmethod
+    def get_zone_or_raise(cls, name: str) -> ForwardZone | ReverseZone:
+        """Get a zone by name, and raise if not found.
+
+        :param name: The name of the zone to get.
+        :returns: The zone object.
+        """
+        zone_t = cls.type_by_name(name)
+        return zone_t.get_by_name_or_raise(name)
+
+    @classmethod
+    def get_zone_and_raise(cls, name: str) -> None:
+        """Get a zone by name, and raise if found.
+
+        :param name: The name of the zone to get.
+        """
+        zone_t = cls.type_by_name(name)
+        return zone_t.get_by_name_and_raise(name)
```

Comment:
Because `ForwardZone` and `ReverseZone` inherit from this class (`Zone`), we can't call these methods `get_by_name*`, since that would shadow the `get_by_name*` methods inherited from `WithName`, hence the different naming scheme here...